### PR TITLE
feat: Add Per-dungeon scaling options

### PR DESCRIPTION
**Pull request by ZhengPeiRu21**
Adds options to allow:

- Setting per-dungeon minimum player count
- Setting per-dungeon inflection value
- Setting per-dungeon boss inflection value

All options are disabled by default.

### CODE REVIEW

**Comment by Helias** [comment]
build is failing, weird error

**Comment by ZhengPeiRu21** [comment]
> build is failing, weird error

I do not think this is related to the PR. I think the CI may be configured incorrect.

**Comment by locus313** [comment]
looks like the workflow has not been updated to the new one yet. pr has been opened to update it:
https://github.com/azerothcore/mod-autobalance/pull/106

**Comment by Helias** [comment]
PR merged, update your branch with master please, to update the build

File: src/AutoBalance.cpp [hunk 1]
```diff
@@ -154,6 +157,78 @@ int GetValidDebugLevel()
...
+    std::stringstream dungeonIdStream;
+
+    dungeonIdStream.str(dungeonIdString);
+    while (std::getline(dungeonIdStream, delimitedValue, ',')) // Process each dungeon ID in the string, delimited by the comma - "," and then space " "
+    {
+        std::string pairOne, pairTwo;
+        std::stringstream dungeonPairStream(delimitedValue);
+        dungeonPairStream>>pairOne>>pairTwo;
+        uint32 dungeonMapId = atoi(pairOne.c_str());
+        float dungeonInflection = atof(pairTwo.c_str());
```
Comment (Helias) [inline src/AutoBalance.cpp:189]: if you can, use the `auto` type when you can instead of the `float` one

Comment (Helias) [reply src/AutoBalance.cpp:189]: same above with uint32 if `atoi` automatically returns `uint32`

File: src/AutoBalance.cpp [hunk 2]
```diff
@@ -303,6 +385,23 @@ class AutoBalance_PlayerScript : public PlayerScript
...
+        {
+            if (DungeonScaleDownMoney)
+            {
+                Map* map = player->GetMap();
+
+                if (map->IsDungeon())
+                {
+                    // Ensure that the players always get the same XP, even when entering the dungeon alone
+                    uint32 maxPlayerCount = ((InstanceMap*)sMapMgr->FindMap(map->GetId(), map->GetInstanceId()))->GetMaxPlayers();
+                    uint32 currentPlayerCount = map->GetPlayersCountExceptGMs();
```
Comment (Helias) [inline src/AutoBalance.cpp:399]: use `auto` as type when you can

Comment (ZhengPeiRu21) [reply src/AutoBalance.cpp:399]: The rest of the module and AzerothCore generally does not use auto. Currently it is consistent with the rest of the module, but I can change to use auto if you would prefer. Would you like me to change these to auto?

Comment (Helias) [reply src/AutoBalance.cpp:399]: yes, it is better
usually we try to use the best practices when we can, some piece of code of AC are surely using `auto` but not all of them unfortunately

Comment (ZhengPeiRu21) [reply src/AutoBalance.cpp:399]: Okay, I will make the change.

**Review by Helias — APPROVED** [review]

**Comment by Helias** [comment]
thank you a lot 🚀

**Pull request merged**